ast/src/parsed/mod.rs

```diff
@@ -336,8 +338,35 @@ impl<R> Children<Expression<R>> for EnumVariant<Expression<R>> {
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema)]
 pub struct TraitImplementation<Expr> {
     pub name: SymbolPath,
+    pub source_ref: SourceRef,
     pub type_scheme: TypeScheme,
-    pub functions: Vec<NamedExpression<Expr>>,
+    pub functions: Vec<NamedExpression<Arc<Expr>>>,
+}
+
+impl<R> TraitImplementation<Expression<R>> {
+    pub fn function_by_name(&self, name: &str) -> Option<&NamedExpression<Arc<Expression<R>>>> {
+        self.functions.iter().find(|f| f.name == name)
+    }
+
+    pub fn function_bodies_mut(&mut self) -> TraitBodyIterMut<R> {
+        TraitBodyIterMut {
+            functions: self.functions.iter_mut(),
+        }
+    }
+}
+
+pub struct TraitBodyIterMut<'a, R> {
```

Comment:
Iterator to be able to obtain mutable references from `Arc<Expr>`. As far as I understand`Arc::get_mut` should not give us problems.

Reply:
Do you actually need a separate struct? `self.functions.iter_mut().map(|named_expr| Arc::get_mut(&mut named_expr.body).unwrap())` does not work?